### PR TITLE
Initial Implementation

### PR DESCRIPTION
**Pull request by azakus**

### CODE REVIEW

File: bin/polymer-css-build [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+  {
+    name: 'help',
+    type: Boolean,
+    description: 'Print this message',
+    alias: 'h'
+  },
+  {
+    name: 'build-for-shady',
+    type: Boolean,
+    description: 'Perform Shady DOM shimming. This will make the build no longer compatible with Shadow DOM'
```
Comment (justinfagnani) [inline bin/polymer-css-build]: would love a link to some docs here, even a section in the README

File: index.js [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+const notStyleMatch = pred.NOT(styleMatch);
+
+const customStyleMatch = pred.AND(
+  pred.hasTagName('style'),
+  pred.hasAttrValue('is', 'custom-style')
+);
+
+const styleIncludeMatch = pred.AND(styleMatch, pred.hasAttr('include'));
+
+const inlineScriptMatch = pred.AND(
```
Comment (justinfagnani) [inline index.js]: I've seen this one before... :)

File: index.js [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+    pred.hasAttrValue('type', 'application/javascript')
+  ),
+  pred.NOT(
+    pred.hasAttr('src')
+  )
+);
+
+const scopeMap = new WeakMap();
+
+function ancestorWalk(node, match) {
```
Comment (justinfagnani) [inline index.js]: TODO: upstream?

File: index.js [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+      );
+      if (parent !== templateContent) {
+        prepend(templateContent, s);
+      }
+    })
+  }
+  return styles;
+}
+
+function getAttributeArray(node, attribute) {
```
Comment (justinfagnani) [inline index.js]: TODO: upstream?

File: index.js [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+    dom5.append(template, content);
+    dom5.append(module, template);
+  } else {
+    styles.forEach(s => {
+      let templateContent = template.childNodes[0];
+      if (!templateContent) {
+        templateContent = dom5.constructors.fragment();
+        dom5.append(template, templateContent);
+      }
+      const parent = ancestorWalk(s, n =>
```
Comment (justinfagnani) [inline index.js]: can you comment what this is doing?

File: index.js [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+
+function prepend(parent, node) {
+  if (parent.childNodes.length > 0) {
+    dom5.insertBefore(parent, parent.childNodes[0], node);
+  } else {
+    dom5.appendChild(parent, node);
+  }
+}
+
+function getDomModuleStyles(module) {
```
Comment (justinfagnani) [inline index.js]: Maybe indicate in the name that this moves styles too.

File: index.js [hunk 7]
```diff
@@ -0,0 +1,335 @@
...
+   * `transform` expects an array of decorated <style> elements
+   *
+   * Decorated <style> elements are ones with `__cssRules` property
+   * with a value of the CSS ast
+   */
+  Polymer.StyleUtil.forEachRule(ast, rule => Polymer.ApplyShim.transformRule(rule));
+}
+
+function getModuleDefinition(module, elements) {
+  for (let i = 0; i < elements.length; i++) {
```
Comment (justinfagnani) [inline index.js]: `for...of`?

File: index.js [hunk 8]
```diff
@@ -0,0 +1,335 @@
...
+    const is = elements[i].is;
+    if (is.toLowerCase() === module) {
+      return elements[i];
+    }
+  }
+  return null;
+}
+
+function getTypeExtends(element) {
```
Comment (justinfagnani) [inline index.js]: Should this be in the ELementDescriptor class?

File: index.js [hunk 9]
```diff
@@ -0,0 +1,335 @@
...
+  const moduleDefinition = getModuleDefinition(scope, elements);
+  // only shim if module is a full polymer element, not just a style module
+  if (!scope || !moduleDefinition) {
+    return;
+  }
+  const ext = getTypeExtends(moduleDefinition);
+  Polymer.StyleTransformer.css(ast, scope, ext);
+}
+
+function addClass(node, className) {
```
Comment (justinfagnani) [inline index.js]: If this is called multiple times, does the element get multiple `style-scope` classes? Consider a name change if it's only called once.

File: index.js [hunk 10]
```diff
@@ -0,0 +1,335 @@
...
+    const template = dom5.query(domModule, pred.hasTagName('template'));
+    // apply scoping to template
+    if (template) {
+      const elements = dom5.queryAll(template, notStyleMatch);
+      elements.forEach(el => addClass(el, scope));
+    }
+  }
+}
+
+module.exports = (paths, options) => {
```
Comment (justinfagnani) [inline index.js]: I would consider not creating ES-incompatible modules. ie, always export a module-type object.

File: index.js (hunk 9 above)
Comment (azakus) [inline index.js]: Done. Classes are now only added if not already found in the list

File: bin/polymer-css-build (hunk 1 above)
Comment (azakus) [inline bin/polymer-css-build]: Done, added big section to README

**Pull request merged**